server/parameters/util/HelpEntry.java

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright (C) 2021 Vaticle
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ */
+
+package com.vaticle.typedb.core.server.parameters.util;
+
+import javax.annotation.Nullable;
+import java.util.List;
+
+import static com.vaticle.typedb.core.common.iterator.Iterators.iterate;
+
+public abstract class HelpEntry {
+
+    final String scopedName;
+    final String description;
+
+    private HelpEntry(String scopedName, String description) {
+        this.scopedName = scopedName;
+        this.description = description;
+    }
+
+    public static abstract class Yaml extends HelpEntry {
+
+        private Yaml(String scopedName, String description) {
+            super(scopedName, description);
+        }
+
+        public static class Grouped extends Yaml {
+
+            private final List<HelpEntry> entries;
+
+            public Grouped(String scopedName, String description, List<HelpEntry> entries) {
+                super(scopedName, description);
+                this.entries = entries;
+            }
+
+            @Override
+            public String toString() {
+                StringBuilder builder = new StringBuilder();
+                if (anyLeafContents()) {
+                    // only print section headers that have a leaf option in them
+                    builder.append(String.format("\n\t### %-50s %s\n", (Option.PREFIX + scopedName + "."), description));
+                }
+                for (HelpEntry menu : entries) {
+                    builder.append(menu.toString());
+                }
+                return builder.toString();
+            }
+
+            private boolean anyLeafContents() {
+                return iterate(entries).anyMatch(content -> content instanceof Simple);
+            }
+        }
+
+    }
+
+    public static class Simple extends HelpEntry {
```

Comment:
don't like this name... can we choose one of the existing ones that isn't related to `Yaml` specifically?
there are some floating around already:
- `KeyValue`
- `Option`
- `Parameter`
what do you think?

Reply:
(I think `Option` is the closest to the domain, it's used for `ArgParser` option helps right?)

Reply:
I don't think we should because then we're naming it based on where it's used.

It is shared equally by `OptionParser`, `YamlParser.KeyValue.Dynamic`, and `YamlParser.KeyValue.Predefined` and therefore the name shouldn't be associated with any of them.

I think `HelpEntry.Simple` is a good name because it represents what it is - a simple entry that has no nesting / grouping.

What do you think?